k2/csrc/cuda/ops.h

```diff
@@ -1,71 +1,142 @@
+// different configuration depending on num_elements and data type.
+constexpr int TILE_DIM = 32;
+constexpr int BLOCK_ROWS = 8;
+
+// TODO(haowen): move the implementations to file `op_inl.h` or
+// `op.cu`(specialized on device and data type)?
+template <typename T>
+__global__ void TransposeKernel(int32_t rows, int32_t cols, const T *input,
+                                T *output) {
+  // TODO(haowen): here we need to handle different type of T to avoid bank
+  // conflicts, the size of cache now is fine for type size with 32bit (e.g.
+  // int32 or float).
+  __shared__ T cache[TILE_DIM][TILE_DIM + 1];
+
+  // input index, in a coalesced manner.
+  int32_t x = threadIdx.x + blockIdx.x * TILE_DIM;
+  int32_t y = threadIdx.y + blockIdx.y * TILE_DIM;
+
+  for (int32_t i = 0; i < TILE_DIM; i += BLOCK_ROWS) {
+    if (x < cols && (y + i) < rows) {
+      cache[threadIdx.y + i][threadIdx.x] = input[(y + i) * cols + x];
+    }
+  }
+
+  __syncthreads();
+
+  // output index, in a coalesced manner
+  x = threadIdx.x + blockIdx.y * TILE_DIM;
+  y = threadIdx.y + blockIdx.x * TILE_DIM;
+  for (int32_t i = 0; i < TILE_DIM; i += BLOCK_ROWS) {
+    if (x < rows && (y + i) < cols) {
+      output[(y + i) * rows + x] = cache[threadIdx.x][threadIdx.y + i];
+    }
+  }
+}
 
 /*
   Transpose a matrix.  Require src.Size0() == dest.Size1() and src.Size1() ==
   dest.Size0().  This is not the only way to transpose a matrix, you can also
-  do: dest = Array2<T>(src.ToTensor().Transpose(0,1)), which will likely call this
-  function
+  do: dest = Array2<T>(src.ToTensor().Transpose(0,1)), which will likely call
+  this function
 
-     @param [in] c   Context to use, must satisfy `c.IsCompatible(src.Context())` and
-                    `c.IsCompatible(dest->Context())`.
+     @param [in] c   Context to use, must satisfy
+  `c.IsCompatible(src.Context())` and `c.IsCompatible(dest->Context())`.
      @param [in] src  Source array to transpose
-     @param [out] dest  Destination array; must satisfy `dest->Size1() == src.Size0()`
-                     and `dest->Size0() == src.Size1()`.  At exit, we'll have
+     @param [out] dest  Destination array; must satisfy `dest->Size1() ==
+  src.Size0()` and `dest->Size0() == src.Size1()`.  At exit, we'll have
                      dest[i,j] == src[j,i].
  */
-template <typename T> void Transpose(ContextPtr &c,
-                                     Array2<T> &src,
-                                     Array2<T> *dest);
-
+template <typename T>
+void Transpose(ContextPtr &c, const Array2<T> &src, Array2<T> *dest) {
+  assert(c.IsCompatible(src.Context());
+  assert(c.IsCompatible(dest->Context());
+  int32_t rows = src.Dim0();
+  int32_t cols = src.Dim1();
+  // TODO(haowen): limit the number of elements?
+  assert(rows == dest->Dim1());
+  assert(cols == dest->Dim0());
+  const T *src_data = src.Data();
+  T *dest_data = dest->Data();
+  // TODO(haowen): replace if-else with template or inheritance?
```

Comment:
@songmeixu we shouldn't be having to do  asserts here, we need an official way to check stuff!!

Reply:
.. and @qindazhu no let's not do template or inheritance.  Just a simple if-else is best I think.

Reply:
OK

Reply:
removed the TODO above